### PR TITLE
[SYSTEMML-1437] Factorization Machines

### PR DESCRIPTION
**Pull request by j143**
Core Script for factorization machines.

### CODE REVIEW

**Comment by j143** [comment]
Hi Mike, there is some issue with one gradient formulation. i.e., gradient of `W` 
cc @dusenberrymw

File: scripts/staging/fm.dml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+ 
+    dw0  = dout %*% (1 / g_w0) ;  # shape has to be (n, 1)
+    
+    # gradient w.r.t. W
+    g_W = X ; # shape (n, d) 
+    
+    dW  =  (1 / t(g_W)) %*% dout; # shape has to be (d, 1)
+    
+    # gradient w.r.t. V
+    g_V = t(X) %*% rowSums(X %*% V) # - (V %*% X^2) ; # shape (d, 1) - (n, k) :\ not correct!
```
Comment (j143) [inline scripts/staging/fm.dml]: @dusenberrymw , second term seems to be a little tricky, where the gradients are not matching.. I tried all possible ways of arranging to get the required dimension, but still something I'm missing out. Please find eq(4) in [paper](https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf)

File: scripts/staging/fm.dml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    return (matrix[double] w0, matrix[double] W, matrix[double] V) {
+   /* 
+    * This function initializes the parameters.
+    * 
+    * Inputs:
+    *  - d: the number of features, is an integer.
+    *  - k: the factorization dimensionality, is an integer.
+    *
+    * Outputs:
+    *  - w0: the global bias, of shape (1,). <- took this as (n, 1)
```
Comment (dusenberrymw) [inline scripts/staging/fm.dml]: This should just be a single value.  `n` is the number of examples in the model.  This is analogous to the bias in the `affine` layer, where `M=1`, and thus the shape is `(1,1)`, or `(1,)` for short.

**Comment by j143** [comment]
Hi @dusenberrymw , I am testing the script right now. So far, the gradient test results are:
```
Grad checking the factorization machines with L2 loss.
 - Grad checking w0.
 - Grad checking W.
 - Grad checking V.
ERROR: Relative error 0.2491806514236036 > 0.01 with 41.07082918688772 analytical vs 24.685
59946109394 numerical, with lossph 107.83819460443586 and lossmh 107.83770089244663
```
If you're free for sometime, please help find where the script needs to modified. I will continue testing till 12:00 AM PST. Thanks.

**Comment by j143** [comment]
@dusenberrymw  at this point, the test results are:
```
Starting grad checks.
---
Grad checking the factorization machines with L2 loss.
 - Grad checking w0.
 - Grad checking W.
 - Grad checking V.
WARNING: Relative error 0.009399454061690514 > 1.0E-4 & <= 0.01 with 100.7248891500922 
analytical vs 102.63637400100832 numerical, with lossph 330.0540992464383 and lossmh 33
0.05204651895826
ERROR: Relative error 0.1795538703049324 > 0.01 with 70.16871310777215 analytical vs 48
.80629070385111 numerical, with lossph 330.05356094497904 and lossmh 330.05258481916496
WARNING: Relative error 0.009399454061690514 > 1.0E-4 & <= 0.01 with 100.7248891500922 
analytical vs 102.63637400100832 numerical, with lossph 330.0540992464383 and lossmh 33
0.05204651895826
```

**Comment by j143** [comment]
With `n = 30` examples, the results are..
```
 - Grad checking w0.
 - Grad checking W.
 - Grad checking V.
ERROR: Relative error 0.9313710389352272 > 0.01 with 921.4261329908262 analytical vs 32.74177614258633 numerical, with lossph 137.27656824966436 and lossmh 137.2759134141415
ERROR: Relative error 0.9331057132016384 > 0.01 with 713.807884601275 analytical vs 24.701012999628343 numerical, with lossph 137.27648784194042 and lossmh 137.27599382168043
ERROR: Relative error 0.9313710389352272 > 0.01 with 921.4261329908262 analytical vs 32.74177614258633 numerical, with lossph 137.27656824966436 and lossmh 137.2759134141415
ERROR: Relative error 0.9331057132016384 > 0.01 with 713.807884601275 analytical vs 24.701012999628343 numerical, with lossph 137.27648784194042 and lossmh 137.27599382168043
ERROR: Relative error 0.9313710389352272 > 0.01 with 921.4261329908262 analytical vs 32.74177614258633 numerical, with lossph 137.27656824966436 and lossmh 137.2759134141415
```

**Comment by j143** [comment]
Hi @dusenberrymw , Happy to say that the FM's are working now!. I've used a small loop, I think this one can be handled easily. 
Thanks.

File: scripts/staging/fm.dml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    # dV = mean(dout) * (t(X) %*% X %*%V) - g_V2
+    xv = X %*% V
+
+    g_V1 = dout[,] * xv[,1]
+
+    for (j in 2:k) {
+      tmp1 = dout[,1] * xv[,k]
+      g_V1 = cbind(g_V1, tmp1)
+    }
+
```
Comment (j143) [inline scripts/staging/fm.dml]: Hi @mboehm7, I have two loops here to multiply corresponding elements, and to retain the structure. I tried some ways.. Can you help me out here.. Thanks.

**Comment by j143** [comment]
Hi @dusenberrymw , Factorization machines are working correctly. May be Merging this part of the work can help boost my confidence.  
Thanks a lot. :)

**Comment by j143** [comment]
Test this Jenkins

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2235/

**Comment by j143** [comment]
I am closing this, for sometime.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/<h2>Failed Tests: <span class='status-failure'>57</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>57</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.codegen/CompressedRowAggregateLargeTest/testCompressedRowAggregateMainDenseConstCP/'><strong>org.apache.sysml.test.integration.functions.codegen.CompressedRowAggregateLargeTest.testCompressedRowAggregateMainDenseConstCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.codegen/CompressedRowAggregateLargeTest/testCompressedRowAggregateMainEmptyRandCP/'><strong>org.apache.sysml.test.integration.functions.codegen.CompressedRowAggregateLargeTest.testCompressedRowAggregateMainEmptyRandCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.codegen/CompressedRowAggregateLargeTest/testCompressedRowAggregateMainEmptyConstCP/'><strong>org.apache.sysml.test.integration.functions.codegen.CompressedRowAggregateLargeTest.testCompressedRowAggregateMainEmptyConstCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.codegen/CompressedRowAggregateLargeTest/testCompressedRowAggregateMainEmptyRand2CP/'><strong>org.apache.sysml.test.integration.functions.codegen.CompressedRowAggregateLargeTest.testCompressedRowAggregateMainEmptyRand2CP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.codegen/CompressedRowAggregateLargeTest/testCompressedRowAggregateMainEmptyRand3CP/'><strong>org.apache.sysml.test.integration.functions.codegen.CompressedRowAggregateLargeTest.testCompressedRowAggregateMainEmptyRand3CP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.codegen/CompressedRowAggregateLargeTest/testCompressedRowAggregateMainSparseConstCP/'><strong>org.apache.sysml.test.integration.functions.codegen.CompressedRowAggregateLargeTest.testCompressedRowAggregateMainSparseConstCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeCompressionTest/testSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeCompressionTest.testSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeCompressionTest/testDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeCompressionTest.testDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeCompressionTest/testEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeCompressionTest.testEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeMatrixVectorMultTest/testSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeMatrixVectorMultTest.testSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeMatrixVectorMultTest/testDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeMatrixVectorMultTest.testDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeMatrixVectorMultTest/testEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeMatrixVectorMultTest.testEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParMatrixVectorMultTest/testSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParMatrixVectorMultTest.testSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParMatrixVectorMultTest/testDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParMatrixVectorMultTest.testDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParMatrixVectorMultTest/testEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParMatrixVectorMultTest.testEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowMaxsEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowMaxsEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testMinSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testMinSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColSumsSqEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColSumsSqEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testSumSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testSumSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowSumsEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowSumsEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowSumsDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowSumsDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColMaxsEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColMaxsEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColSumsSqSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColSumsSqSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColSumsEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColSumsEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColSumsSqDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColSumsSqDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testSumDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testSumDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testMinEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testMinEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testSumSqEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testSumSqEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testMaxEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testMaxEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowMaxsDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowMaxsDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testMaxDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testMaxDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColMinsDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColMinsDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColMinsSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColMinsSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColSumsSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColSumsSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowSumsSqSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowSumsSqSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColSumsDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColSumsDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testSumSqDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testSumSqDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColMaxsSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColMaxsSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testMaxSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testMaxSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testSumSqSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testSumSqSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowMinsEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowMinsEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowSumsSqDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowSumsSqDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowMinsSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowMinsSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColMaxsDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColMaxsDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowSumsSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowSumsSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testColMinsEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testColMinsEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testMinDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testMinDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowMinsDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowMinsDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowSumsSqEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowSumsSqEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testRowMaxsSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testRowMaxsSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeParUnaryAggregateTest/testSumEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeParUnaryAggregateTest.testSumEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeVectorMatrixMultTest/testSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeVectorMatrixMultTest.testSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeVectorMatrixMultTest/testDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeVectorMatrixMultTest.testDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/LargeVectorMatrixMultTest/testEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.LargeVectorMatrixMultTest.testEmptyCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/ParCompressionTest/testSparseConstDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.ParCompressionTest.testSparseConstDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/ParCompressionTest/testDenseConstantDataCompression/'><strong>org.apache.sysml.test.integration.functions.compress.ParCompressionTest.testDenseConstantDataCompression</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2238/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.compress/ParCompressionTest/testEmptyCompression/'><strong>org.apache.sysml.test.integration.functions.compress.ParCompressionTest.testEmptyCompression</strong></a></li></ul>

**Comment by j143** [comment]
Simply, it can take < 6 mins to review this PR,

1. has model equation, 
fm_model
`  Line 44: out = (X %*% W) + (0.5 * rowSums((X %*% V)^2 - (X^2 %*% V^2)) ) + w0; `

2. gradients
image

`Line 72:  w0 -> dw0  = colSums(dout) ;`

`Line 78:  dW  =  t(g_W) %*% dout;`

3. gradient of V (can take only 3 minutes to verify)
I've used a little loop, but a matrix operation can be used instead here. 
`Line  106: dV = (t(X) %*% g_V1) - g_V2`


Thanks,
Janardhan

**Comment by j143** [comment]
Hi Mike, as there is nobody from whom I could take help from. I cannot work on Factorization Machines any more. This is all I can contribute. 

Thanks a lot for all the help,
Janardhan

cc @dusenberrymw

**Comment by j143** [comment]
sorry.

**Comment by dusenberrymw** [comment]
@j143 Can you reopen this?

**Comment by j143** [comment]
Thanks @dusenberrymw , I believe the the loop that I've used can be vectorized easily. 
1. Vectorizing the loop, i.e., multiplying each column of one matrix with corresponding elements of a column matrix.
2. Now, gradient wrt. the input (i.e., X) will be very easier after that.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2250/

**Comment by j143** [comment]
If the community can give the confidence that this jira(1437) can be reviewed with high priority (i.e., with in 2 days of pr submission & merge ASAP), I will try submitting other scripts of `fm`.

**Comment by dusenberrymw** [comment]
@j143 Agreed, vectorizing the loop will be key to ensure that this layer is performant.  That being said, we can merge this in now, and keep the vectorization as a follow-up TODO item.  Also, I think it would be good to move the `fm` file into the `nn/layers` folder since it follows the `nn` API.

**Comment by j143** [comment]
Thanks for the support, Mike. I will run tests on the other scripts from now on. 
cc @dusenberrymw

**Comment by krishnakalyan3** [comment]
@j143 could you also update the documentation for FM or create a JIRA to do that?.

**Comment by j143** [comment]
Hi, I decided not to work on this, till next three months. Thanks.

**Comment by dusenberrymw** [comment]
Merged!  Thanks, @j143 for the hard work!  I created a few other follow-up PRs, in addition to the regression and classification scripts.  I would encourage you to reopen those PRs.

**Pull request closed without merging**